classes/core/PKPPageRouter.php

```diff
@@ -500,40 +500,44 @@ private function _getLocaleForUrl(PKPRequest $request, ?Context $context, array
      */
     private function _setLocale(PKPRequest $request, ?string $setLocale): void
     {
-        $contextPath = $this->_getRequestedUrlParts(['Core', 'getContextPath'], $request);
-        $urlLocale = $this->_getRequestedUrlParts(['Core', 'getLocalization'], $request);
+        $contextPath = $this->_getRequestedUrlParts(Core::getContextPath(...), $request);
+        $urlLocale = $this->_getRequestedUrlParts(Core::getLocalization(...), $request);
         $multiLingual = count($this->_getContextAndLocales($request, $contextPath)[1]) > 1;
-
-        if (!$multiLingual && !$urlLocale && !$setLocale || $multiLingual && !$setLocale && $urlLocale === Locale::getLocale()) {
+        // Quit if there's no new locale to be set and the request URL is already well-formed
+        if (!$setLocale && ($multiLingual ? $urlLocale === Locale::getLocale() : !$urlLocale)) {
             return;
         }
 
-        $sessionLocale = (function (string $l) use ($request): string {
-            $session = $request->getSession();
-            if (Locale::isSupported($l) && $l !== $session->get('currentLocale')) {
-                $session->put('currentLocale', $l);
-                $request->setCookieVar('currentLocale', $l);
-            }
-            // In case session current locale has been set to non-supported locale, or is null, somewhere else
-            if (!Locale::isSupported($session->get('currentLocale') ?? '')) {
-                $session->put('currentLocale', Locale::getLocale());
-                $request->setCookieVar('currentLocale', Locale::getLocale());
-            }
-            return $session->get('currentLocale');
-        })($setLocale ?? $urlLocale);
+        $session = $request->getSession();
+        $currentLocale = $session->get('currentLocale') ?? '';
+        if (!Locale::isSupported($currentLocale ?? '')) {
+            $currentLocale = null;
+        }
+
+        $newLocale = $setLocale ?? $urlLocale;
+        if (!Locale::isSupported($newLocale)) {
+            $newLocale = $currentLocale ?? Locale::getLocale();
+        }
+
+        if ($newLocale !== $currentLocale) {
+            $session->put('currentLocale', $newLocale);
+            $request->setCookieVar('currentLocale', $newLocale);
+        }
 
+        // Do not permit basic auth strings (user:password@url) in source parameter for redirects
         if (preg_match('#^/\w#', $source = str_replace('@', '', $request->getUserVar('source') ?? ''))) {
             $request->redirectUrl($source);
         }
 
+        $newUrlLocale = $multiLingual ? "/{$newLocale}" : '';
         $indexUrl = $this->getIndexUrl($request);
-        $uri = preg_replace("#^{$indexUrl}#", '', $setLocale ? ($_SERVER['HTTP_REFERER'] ?? '') : $request->getCompleteUrl(), 1);
```

Comment:
On the line 525 there's a `str_replace('@', '', $url)`, was it used to do some kind of abuse (redirect user to another system and pass credentials)?

If yes, then the `HTTP_REFERER` here should receive the same treatment, as it's also unsafe, depending on what was the issue, perhaps we should check if the destiny URL belongs to us (e.g. same domain/base path).

Reply:
This was added here: https://github.com/pkp/pkp-lib/issues/10478.
So maybe @asmecher would know best.

Reply:
I think my brain is not functioning well enough to see/think what could happen with HTTP_REFERER :sweat_smile: 
It could be a fully different web page from where the request with setLocale is coming and we would then try to replace contexPath+oldLocale with contextPath+newLocale in that URL, add that string then to $indexUrl (which is checking after our base URL) and redirect to it :thinking: 
And I just cannot figure out what 'bad' could happen thereby... so that we take good care of it... :thinking:

Reply:
This won't be relevant after my update, as now I'm checking if the `preg_replace()` succeeded (and redirecting to `/index/{newLocale}` if not). But it might be relevant for the `stable-3_3_0`, see: https://github.com/pkp/pkp-lib/blob/0d026553661e8a5137e1fc727cf3728eafea1327/pages/user/PKPUserHandler.inc.php#L50-L52

Reply:
Whatever, I'll mark it as resolved, in order for this to be risky, we should be passing extra information back to the URL (e.g. credentials).